src/xml/builder.cr

```diff
@@ -180,14 +183,17 @@ struct XML::Builder
 
   # Emits the start of a `CDATA` section, invokes the block
   # and then emits the end of the `CDATA` section.
-  def cdata
+  #
+  # NOTE: CDATA end sequences written within the block
+  # need to be escaped manually.
+  def cdata(&)
     start_cdata
     yield.tap { end_cdata }
   end
 
-  # Emits a `CDATA` section.
+  # Emits a `CDATA` section.  Escapes nested CDATA end sequences.
   def cdata(text : String) : Nil
-    call WriteCDATA, string_to_unsafe(text)
+    call WriteCDATA, string_to_unsafe(text.gsub(CDATA_END, CDATA_ESCAPE))
```

Comment:
I think ideally `gsub!` would make more sense here to prevent needing to allocate another string.  But unfortunately that method doesn't exist :/

Reply:
Yes, strings are immutable. You wouldn't want to modify the caller's value anyways.

Reply:
I guess potentially faster and maybe we allocation free could be splitting on the end marker and writing each part separately? Nothing I would require here though.

Reply:
Strings are immutable. There are no mutating methods on `String` at all. And it would be quite unexpected that a string is changed in the call scope after it is passed to this method.

Reply:
Oops, somehow the comment thread was not updated while the review view was open.